src/primitives/transaction.h

```diff
@@ -170,6 +248,124 @@ class OutputDescription
+        READWRITE(valueBalanceSapling);
+        READWRITE(anchorSapling);
+        if (ser_action.ForRead()) {
+            for (auto &spend : vSpendsSapling) {
+                libzcash::GrothProof zkproof;
+                READWRITE(zkproof);
+                vSpendProofsSapling.push_back(zkproof);
+            }
+            for (auto &spend : vSpendsSapling) {
+                SpendDescription::spend_auth_sig_t spendAuthSig;
+                READWRITE(spendAuthSig);
+                vSpendAuthSigSapling.push_back(spendAuthSig);
+            }
+            for (auto &output : vOutputsSapling) {
+                libzcash::GrothProof zkproof;
+                READWRITE(zkproof);
+                vOutputProofsSapling.push_back(zkproof);
+            }
+        } else {
+            for (auto &zkproof : vSpendProofsSapling) {
+                READWRITE(zkproof);
+            }
+            for (auto &spendAuthSig : vSpendAuthSigSapling) {
+                READWRITE(spendAuthSig);
+            }
+            for (auto &zkproof : vOutputProofsSapling) {
+                READWRITE(zkproof);
+            }
+        }
+        READWRITE(bindingSigSapling);
+    }
+
+    std::vector<SpendDescription> GetV4ShieldedSpend();
+    std::vector<OutputDescription> GetV4ShieldedOutput();
+};
+
+/**
+ * The Orchard component of a transaction.
+ */
+class OrchardBundle
+{
+private:
+    /// An optional Orchard bundle (with `nullptr` corresponding to `None`).
+    /// Memory is allocated by Rust.
+    std::unique_ptr<OrchardBundlePtr, decltype(&orchard_bundle_free)> inner;
+
+public:
+    OrchardBundle() : inner(nullptr, orchard_bundle_free) {}
+
+    OrchardBundle(OrchardBundle&& bundle) : inner(std::move(bundle.inner)) {}
+    OrchardBundle(const OrchardBundle& bundle) : inner(orchard_bundle_clone(bundle.inner.get()), orchard_bundle_free) {}
+    OrchardBundle& operator=(OrchardBundle&& bundle)
+    {
+        if (this != &bundle) {
+            inner = std::move(bundle.inner);
+        }
+        return *this;
+    }
+    OrchardBundle& operator=(const OrchardBundle& bundle)
+    {
+        if (this != &bundle) {
+            inner.reset(orchard_bundle_clone(bundle.inner.get()));
+        }
+        return *this;
+    }
+
+    template<typename Stream>
+    void Serialize(Stream& s) const {
+        unsigned char* buf;
+        size_t bufCapacity;
+        size_t orchardBytesLen = orchard_bundle_serialize(inner.get(), buf, &bufCapacity);
+        s.write(reinterpret_cast<const char*>(buf), orchardBytesLen);
+        orchard_buffer_free(buf, bufCapacity);
+    }
+
+    template<typename Stream>
+    void Unserialize(Stream& s) {
+        size_t orchardBytesRead = orchard_bundle_parse(
+            reinterpret_cast<const unsigned char*>(s.data()), s.size(), inner.get());
+        if (orchardBytesRead == 0) {
+            throw std::ios_base::failure("Failed to parse v5 Orchard bundle");
+        }
+        s.ignore(orchardBytesRead);
```

Comment:
This approach would work if we only ever parsed transactions from `CDataStream`s, but unfortunately we also parse them from `CDataStream`-like wrappers around file handles, for which we can't use `.data()` and `.size()`. I'll try working around this by specializing those cases to understand the bare minimum of Orchard bundle-parsing logic, to know how much data to read and buffer.

Reply:
I instead wrote a C++ stream wrapper so we could use them directly in Rust.